### PR TITLE
nec-usbatermfw: add tool for NEC "USBATERM" format

### PR DESCRIPTION
**Pull request by musashino205**
Add a tool for NEC Aterm devices using "USBATERM" format. This format is mainly used on the devices based on NetBSD, such as WR8750N, WG600HP and WG1400HP.

A workaround in write_blockheader() is needed to resolve an unknown bug when generating an image if built with gcc 9.4.0
(Ubuntu 9.4.0-1ubuntu1~20.04.2) and cmake.

example:

index 1 and 4 are fixed data and checksum of them shouldn't be changed

- with the workaround

  ```
  nec-usbatermfw \
        openwrt-ath79-tiny-nec_wr9500n-initramfs-factory.bin.new \
        -t H033 \
        -f 0x0003 -d loader-nec_wr9500n.bin \
        -d openwrt-ath79-tiny-nec_wr9500n-initramfs-factory.bin
          Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0003, 0x00000a14,     0x80040000,     0x80040000,     0x00001526
  3:      0x0002, 0x0054df20,     0x80040000,     0x80040000,     0x0000c41f
  4:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
  ```

- without the workaround

  ```
  nec-usbatermfw \
        openwrt-ath79-tiny-nec_wr9500n-initramfs-factory.bin.new \
        -t H033 \
        -f 0x0003 -d loader-nec_wr9500n.bin \
        -d openwrt-ath79-tiny-nec_wr9500n-initramfs-factory.bin
          Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0003, 0x00000a14,     0x80040000,     0x80040000,     0x00009528
  3:      0x0002, 0x0054df20,     0x80040000,     0x80040000,     0x00004422
  4:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x00005002
  ```

### CODE REVIEW

**Comment by musashino205** [comment]
@svanheule @hauke Can you handle this? The official buildbot has failed to build ath79/tiny due to missing **nec-usbatermfw** tool.

File: src/nec-usbatermfw.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		return -EINVAL;
+	}
+
+	outbin = fopen(argv[1], "w+");
+	if (!outbin) {
+		fprintf(stderr, "Couldn't open %s\n", argv[1]);
+		return -EACCES;
+	}
+
+	while((c = getopt(argc, argv, "f:a:e:d:t:")) != -1) {
```
Comment (hauke) [inline src/nec-usbatermfw.c:227]: These two `getopt` loops are strange. 

I would prefer if you put this into one `getopt` loop.

File: src/nec-usbatermfw.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+	return padlen;
+}
+
+static int write_blockheader(uint32_t flags,
+			     uint32_t loadaddr, uint32_t entryp,
+			     size_t datalen, int padlen, uint32_t cksum)
+{
+	struct header hdr;
+	void *cur = &hdr;
```
Comment (hauke) [inline src/nec-usbatermfw.c:113]: Why not make this a `uint16_t` pointer?
```suggestion
	uint16_t *cur = (uint16_t *)&hdr;
```
Then you need less casts in the for loop.

**Comment by hauke** [comment]
Could you share the broken binary comiled on ubuntu 20.04 please. I would like to have a look at it.

**Comment by musashino205** [comment]
[nec-usbatermfw.zip](https://github.com/user-attachments/files/16478225/nec-usbatermfw.zip)

compiled on:

- OS

  ```
  $ lsb_release -a
  No LSB modules are available.
  Distributor ID: Ubuntu
  Description:    Ubuntu 20.04.6 LTS
  Release:        20.04
  Codename:       focal
  ```

- gcc (/usr/bin/cc)

  ```
  $ ls -al staging_dir/host/bin/gcc
  lrwxrwxrwx 1 tofu tofu 11 Aug  3 14:13 staging_dir/host/bin/gcc -> /usr/bin/cc
  $ staging_dir/host/bin/gcc --version
  gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
  Copyright (C) 2019 Free Software Foundation, Inc.
  This is free software; see the source for copying conditions.  There is NO
  warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

  ```

File: src/nec-usbatermfw.c (hunk 1 above)
Comment (musashino205) [reply src/nec-usbatermfw.c:227]: OK, I'll update...

**Comment by musashino205** [comment]
UPDATE:

- use `uint16_t *` for `cur` pointer instead of `void *`
- merge the 2x `getopt()` loops

**Comment by hauke** [comment]
Thank you for the binary.

My arch Linux generates this code for the loop:
```
  do {
    uVar1 = *cur;
    cur = (uint16_t *)((long)cur + 2);
    cksum = cksum + uVar1;
  } while ((long)cur - (long)hdr < 0x17);
```

Your Ubuntu 20.04 generates this code:
```
  do {
    uVar1 = *cur;
    cur = (uint16_t *)((long)cur + 2);
    cksum = cksum + uVar1;
  } while ((long)cur - (long)hdr < 0x18);
```

I do not understand why gcc 9.4.0 does an extra loop here.

**Comment by svanheule** [comment]
Since you want to compare the length of the header in bytes anyway, does it help when you cast to `uintptr_t` instead on Ubuntu 20.04 + gcc 9.4.0?

```
for (cur += 2; (uintptr_t) cur - (uintptr_t) &hdr < BLKHDR_LEN; cur++) {...}
```

**Comment by musashino205** [comment]
> Since you want to compare the length of the header in bytes anyway, does it help when you cast to uintptr_t instead on Ubuntu 20.04 + gcc 9.4.0?
>
> ```
> for (cur += 2; (uintptr_t) cur - (uintptr_t) &hdr < BLKHDR_LEN; cur++) {...}
> ```

Thanks, but it doesn't help... 😭

I tested under some conditions:

(Note: the same loader binary is used)

- compile using `cc` directly without the workaround (`cc -o nec-usbatermfw nec-usbatermfw.c`)

  ```
  $ ./nec-usbatermfw test.bin -t H033 -d build_dir/target-mips_24kc_musl/linux-ath79_generic/loader-nec_wg1400hp.bin
        Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0002, 0x000009f4,     0x80040000,     0x80040000,     0x0000bf57
  3:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
  ```

- compile using `cc` and `cmake` without the workaround (`make tools/firmware-utils/compile`)

  ```
  $ staging_dir/host/bin/nec-usbatermfw test.bin -t H033 -d build_dir/target-mips_24kc_musl/linux-ath79_generic/loader-nec_wg1400hp.bin
          Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0002, 0x000009f4,     0x80040000,     0x80040000,     0x00003f5a <-- bad checksum
  3:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
  ```

- compile using `cc`, `cmake` and `uintptr_t` without the workaround (`make tools/firmware-utils/compile`)

  ```
  $ staging_dir/host/bin/nec-usbatermfw test.bin -t H033 -d build_dir/target-mips_24kc_musl/linux-ath79_generic/loader-nec_wg1400hp.bin
          Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0002, 0x000009f4,     0x80040000,     0x80040000,     0x00003f5a <-- bad checksum
  3:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
  ```

- compile using `cc`, `cmake` with the workaround (`make tools/firmware-utils/compile`)

  ```
  $ staging_dir/host/bin/nec-usbatermfw test.bin -t H033 -d build_dir/target-mips_24kc_musl/linux-ath79_generic/loader-nec_wg1400hp.bin
          Flags   Total Len.      Load Addr       Entry Point     Checksum
  1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
  2:      0x0002, 0x000009f4,     0x80040000,     0x80040000,     0x0000bf57
  3:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
  ```

**Comment by svanheule** [comment]
One last attempt then: writing the loop differently.

```c
uint16_t *cur = (uint16_t *) &hdr;
/* BLKHDR_LEN is a multiple of sizeof(uint16_t), so rounding doesn't matter */
uint16_t * const end = cur + (BLKHDR_LEN / sizeof(uint16_t));

for (cur += 2; cur != end; cur++) {
    chksum += *cur;
}
```

`cc` also seems to generate code matching this behaviour when speed optimisations are used.

**Review by svanheule — COMMENTED** [review]
Some minor comments regarding the endianess macros. Nothing deal breaking however.

**Comment by musashino205** [comment]
UPDATE:

- use `uint32_t` instead of `unsigned int` for `*val` parameter on `strtou32()`
- use `*toh`/`hto*` endianness functions instead of macros (`HOST_TO_*`)
- replace to `le16toh` from `HOST_TO_LE16` in `write_data()`

---

> One last attempt then: writing the loop differently.

I tried, but it doesn't resolve the issue.

```
$ staging_dir/host/bin/nec-usbatermfw test.bin -t H033 -d build_dir/target-mips_24kc_musl/linux-ath79_generic/loader-nec_wg1400hp.bin
        Flags   Total Len.      Load Addr       Entry Point     Checksum
1:      0x0000, 0x0000003e,     0x00000000,     0x00000000,     0x00001988
2:      0x0002, 0x000009f4,     0x80040000,     0x80040000,     0x00003f58
3:      0x0001, 0x00000018,     0x00000000,     0x00000000,     0x0000cfff
```

File: src/nec-usbatermfw.c [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+	hdr.flags = htobe32(flags);
+	hdr.totlen = htobe32(BLKHDR_LEN + datalen);
+	hdr.hdrlen = htobe32(BLKHDR_LEN);
+	hdr.cksum = 0;
+	hdr.loadaddr = htobe32(loadaddr);
+	hdr.entryp = htobe32(entryp);
+
+	for (cur += 2; cur - (uint16_t *)&hdr < BLKHDR_LEN / 2; cur++) {
+		cksum += htole16(*cur);
```
Comment (hauke) [inline src/nec-usbatermfw.c:115]: Isn't this:
```suggestion
		cksum += be16toh(*cur);
```

Comment (musashino205) [reply src/nec-usbatermfw.c:115]: No, it should be `htole16` (or `le16toh`).

```
IPL:HARD-RESET
memory test ... ok
flinstall OK

boot version: 1.0.0

PKG "Orthos-HP"
CPU "Atheros/AR9344"
Revision 212.1 CLOCK 560.0 MHz
HEAP MEMORY 262144 bytes
Boot:InitFilesystem
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
post start
DRAM Test ..... PASS
Timer test .... PASS
FLASH CHKSUM .. PASS
LAN Test ..... 
LoopBack count 0 .. 
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
active ch:1
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
PASS
Package TYPE = 0x01
post success

welcome to boot console
Password:
boot> tftpd
tftpd start 192.168.0.1
boot> start tftp load openwrt-ath79-tiny-nec_wg600hp-i         <-- be16toh()
end tftp load length = 5578496
start memory load ... 
load error [5]

boot> start tftp load openwrt-ath79-tiny-nec_wg600hp-i         <-- htole16()
end tftp load length = 5578496
start memory load ... 
memory load complete
  begin  : 0x80040000
  length : 5575740
  startup: 0x80040000

```

**Pull request merged**